src/sentry/sentry_metrics/indexer/ratelimiters.py

```diff
@@ -0,0 +1,169 @@
+    GrantedQuota,
+    Quota,
+    RedisSlidingWindowRateLimiter,
+    RequestedQuota,
+    Timestamp,
+)
+from sentry.sentry_metrics.configuration import UseCaseKey, get_ingest_config
+from sentry.sentry_metrics.indexer.base import FetchType, FetchTypeExt, KeyCollection, KeyResult
+from sentry.utils import metrics
+
+OrgId = int
+
+
+def _build_quota_key(use_case_id: UseCaseKey, org_id: Optional[OrgId] = None) -> str:
+    use_case_str = {
+        UseCaseKey.PERFORMANCE: "performance",
+        UseCaseKey.RELEASE_HEALTH: "releasehealth",
+    }[use_case_id]
+
+    if org_id is not None:
+        return f"metrics-indexer-{use_case_str}-org-{org_id}"
+    else:
+        return f"metrics-indexer-{use_case_str}-global"
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quotas")
+def _construct_quotas(use_case_id: UseCaseKey) -> Sequence[Quota]:
+    """
+    Construct write limit's quotas based on current sentry options.
+
+    This value can potentially cached globally as long as it is invalidated
+    when sentry.options are.
+    """
+    if use_case_id == UseCaseKey.PERFORMANCE:
+        return [
+            Quota(prefix_override=_build_quota_key(UseCaseKey.PERFORMANCE, None), **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.performance.global")
+        ] + [
+            Quota(prefix_override=None, **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.performance.per-org")
+        ]
+    elif use_case_id == UseCaseKey.RELEASE_HEALTH:
+        return [
+            Quota(prefix_override=_build_quota_key(UseCaseKey.RELEASE_HEALTH, None), **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.global")
+        ] + [
+            Quota(prefix_override=None, **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.per-org")
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quota_requests")
+def _construct_quota_requests(
+    use_case_id: UseCaseKey, keys: KeyCollection
+) -> Tuple[Sequence[OrgId], Sequence[RequestedQuota]]:
+    org_ids = []
+    requests = []
+    quotas = _construct_quotas(use_case_id)
+
+    if quotas:
+        for org_id, strings in keys.mapping.items():
+            org_ids.append(org_id)
+            requests.append(
+                RequestedQuota(
+                    prefix=_build_quota_key(use_case_id, org_id),
+                    requested=len(strings),
+                    quotas=quotas,
+                )
+            )
+
+    return org_ids, requests
+
+
+@dataclasses.dataclass
+class RateLimitState:
+    use_case_id: UseCaseKey
+    requests: Sequence[RequestedQuota]
+    grants: Sequence[GrantedQuota]
+    timestamp: Timestamp
+
+
+@dataclasses.dataclass
+class DroppedString:
+    key_result: KeyResult
+    fetch_type: FetchType
+    fetch_type_ext: FetchTypeExt
+
+
+class WritesLimiter:
```

Comment:
Why not making this a context manager? 
Then you would not have to give back the state to the code that writes on postgres and relying on that to pass it back to the rate limtier to `apply_write_limits`.

Example is how we did it in snuba 
https://github.com/getsentry/snuba/blob/master/snuba/state/rate_limit.py#L94

You can yield the list of strings to write and those to drop. 
Personally I would make it a class and not a generator, but this is just personal preference.

Reply:
done. I made the `state` itself a context manager